Assets/Scripts/Managers/BattleQueue.cs

```diff
@@ -92,7 +106,105 @@ void RenderBQ()
     //Gives BattleQueue ownership of the lifetime of the Dequeue coroutine.
     public void BeginDequeue()
     {
-        StartCoroutine(Dequeue());
+        // StartCoroutine(Dequeue());
+
+        StartCoroutine(DequeueWrappers());
+    }
+
+    // Begins the dequeueing process. 
+    // REQUIRES: An appropriate call. Note that this can be called even if the number of elements in the actionQueue is 0. Invariant array index 0 has largest speed. 
+    // MODIFIES: the actionQueue is progressively emptied until it is empty. 
+    public IEnumerator DequeueWrappers()
+    {
+        List<Wrapper> array = wrapperArray.GetWrappers();
+        bool beganFighting = false;
+        if (!(array.Count == 0))
+        {
+            CombatManager.Instance.GameState = GameState.FIGHTING;
+            beganFighting = true;
+        }
+
+        int i = 0;
+
+        while (!(array.Count == 0))
+        {
+            if (i == 0)
+            {
+                wrapperArray.DisplayWrapperArray();
+                i = 1;
+            }
+
+
+            Wrapper e = array[0];
+
+            ActionClass action = e.ReturnHighest(); // accounts for half-wrappers too 
+
+            if (action == null) 
+            {
+                throw new Exception("Empty Action Returned.");
+            }
+
+            // %%% could be optimised for logic
+            // removing from the Wrapper based on type
+            if (action.IsPlayedByPlayer()) // debugging 
+            {
+                Debug.Log("A player action of speed x is removed: " + action.Speed);
+            }
+            else
+            {
+                Debug.Log("An enemy action of speed x is removed: " + action.Speed);
+            }
+
+            if (e.HighestSpeed == -1 || (e.PlayerAction == null && e.EnemyAction == null))
+            {
+                if (e.EnemyAction != null || e.PlayerAction != null)
+                {
+                    throw new Exception("Wrapper Invariant not Upheld"); // $$$ what are the implications for the coroutine and the exception? 
+                }
+                array.Remove(e); // wrapper is now worthless
+            }
+            else if (e.PlayerAction != null || e.EnemyAction != null)// the wrapper has a component left 
+            {
+                if (action.IsPlayedByPlayer())
+                {
+                    // insert enemy action 
+                    // wrapperArray.InsertEnemyActionIntoWrappers(e.EnemyAction);
+                    wrapperArray.GetWrappers().Remove(e); 
+                    wrapperArray.InsertEnemyActionIntoWrappers(e.EnemyAction);
+                }
+                else
+                {
+                    // insert player action
+                    // wrapperArray.InsertPlayerActionIntoWrappers(e.PlayerAction);
+                    wrapperArray.GetWrappers().Remove(e); 
+                    wrapperArray.InsertPlayerActionIntoWrappers(e.PlayerAction);
```

Comment:
There Really should only be a single insert method that then does all this logic for you